File: scripts/key_derivate.py
import key_generate

from cryptography.hazmat.backends import default_backend
from cryptography.hazmat.primitives import hashes
from cryptography.hazmat.primitives.asymmetric import ec
from cryptography.hazmat.primitives.kdf.pbkdf2 import PBKDF2HMAC

debug: bool = False


def derive_key(private_key: ec.EllipticCurvePrivateKey, public_key: ec.EllipticCurvePublicKey) -> bytes:
    shared_secret = private_key.exchange(ec.ECDH(), public_key)

    return shared_secret


if __name__ == '__main__':
    alice_private_key, alice_public_key = key_generate.generate_key_pair()
    bob_private_key, bob_public_key = key_generate.generate_key_pair()

    alice_derived_key = derive_key(alice_private_key, bob_public_key)
    bob_derived_key = derive_key(bob_private_key, alice_public_key)

    if debug:
        print("alice_derived_key: ", alice_derived_key)
        print("bob_derived_key: ", bob_derived_key)
        print()

    assert alice_derived_key == bob_derived_key

    print("All tests passed.")

File: scripts/keyDerivate.js
const crypto = require('crypto');

const debug = false;

function deriveKey(privateKey, publicKey) {

  const sharedSecret = crypto.diffieHellman({
    privateKey,
    publicKey,
    privateKeyEncoding: {
      format: 'pem',
      type: 'pkcs8'
    },
    publicKeyEncoding: {
      format: 'pem',
      type: 'spki'
    }
  });

  return sharedSecret;
}
module.exports = { deriveKey };

if (require.main === module) {
  const keyGenerate = require('./keyGenerate');

  const aliceKeyPair = keyGenerate.generateKeyPair();
  const bobKeyPair = keyGenerate.generateKeyPair();

  if (debug) {
    console.log('Alice private key: ', aliceKeyPair.privateKey);
    console.log('Alice public key: ', aliceKeyPair.publicKey);
    console.log('Bob private key: ', bobKeyPair.privateKey);
    console.log('Bob public key: ', bobKeyPair.publicKey);
    console.log()
  }

  const sharedSecret1 = deriveKey(aliceKeyPair.privateKey, bobKeyPair.publicKey);
  const sharedSecret2 = deriveKey(bobKeyPair.privateKey, aliceKeyPair.publicKey);

  if (debug) {
    console.log('Shared secret 1: ', sharedSecret1);
    console.log('Shared secret 2: ', sharedSecret2);
  }

  console.assert(sharedSecret1, sharedSecret2);

  console.log("All tests passed!");
}


